instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb

```diff
@@ -0,0 +1,277 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require_relative '../util'
+require 'opentelemetry/trace/status'
+
+module OpenTelemetry
+  module Instrumentation
+    module Rack
+      module Middlewares
+        # OTel Rack Event Handler
+        #
+        # This seeds the root context for this service with the server span as the `current_span`
+        # allowing for callers later in the stack to reference it using {OpenTelemetry::Trace.current_span}
+        #
+        # It also registers the server span in a context dedicated to this instrumentation that users may look up
+        # using {OpenTelemetry::Instrumentation::Rack.current_span}, which makes it possible for users to mutate the span,
+        # e.g. add events or update the span name like in the {ActionPack} instrumentation.
+        #
+        # @example Rack App Using BodyProxy
+        #   GLOBAL_LOGGER = Logger.new($stderr)
+        #   APP_TRACER = OpenTelemetry.tracer_provider.tracer('my-app', '1.0.0')
+        #
+        #   Rack::Builder.new do
+        #     use Rack::Events, [OpenTelemetry::Instrumentation::Rack::Middlewares::EventHandler.new]
+        #     run lambda { |_arg|
+        #       APP_TRACER.in_span('hello-world') do |_span|
+        #         body = Rack::BodyProxy.new(['hello world!']) do
+        #           rack_span = OpenTelemetry::Instrumentation::Rack.current_span
+        #           GLOBAL_LOGGER.info("otel.trace_id=#{rack_span.context.hex_trace_id} otel.span_id=#{rack_span.context.hex_span_id}")
+        #         end
+        #         [200, { 'Content-Type' => 'text/plain' }, body]
+        #       end
+        #     }
+        #   end
+        #
+        # @see Rack::Events
+        # @see OpenTelemetry::Instrumentation::Rack.current_span
+        class EventHandler
+          include ::Rack::Events::Abstract
+
+          TOKENS_KEY = 'otel.context.tokens'
+          GOOD_HTTP_STATUSES = (100..499).freeze
+
+          # Creates a server span for this current request using the incoming parent context
+          # and registers them as the {current_span}
+          #
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] This is nil in practice
+          # @return [void]
+          def on_start(request, _)
+            return if untraced_request?(request.env)
+
+            extracted_context = extract_remote_context(request)
+            if record_frontend_span?
+              request_start_time = OpenTelemetry::Instrumentation::Rack::Util::QueueTime.get_request_start(request.env)
+              unless request_start_time.nil?
+                frontend_span = tracer.start_span(
```

Comment:
@fbogsany @mwear Am I correct in that the frontend span is here for parity with `dd-trace-rb`? Do we have any insight into how popular this feature is? Would it be ok to consider dropping it?

https://github.com/DataDog/dd-trace-rb/blob/2b163dd91c8d349ceb92370c16f4a7be4a691c35/docs/GettingStarted.md#http-request-queuing

I see you all discussed how this should work early: https://github.com/open-telemetry/opentelemetry-specification/issues/51

AFAICT representing proxy spans remains unresolved: https://github.com/open-telemetry/opentelemetry-specification/issues/51

But it looks like we took this approach:
https://github.com/open-telemetry/opentelemetry-specification/pull/337#issuecomment-549567281

One thing I noticed about porting this functionality to use the events API is that the `http.proxy` span does not include semantic attributes although it is the `server` span. Is that intentional? 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb#L109

Reply:
We discussed this at the SIG today. 

@robbkidd and I are going to swarm on resolving proxy spans with the HTTP WG but we will be dropping this feature.

Endusers may add `X-Request-Start` or `X-Queue-Time` via `allowed_request_headers` and process the span attributes in their vendor backends.

Reply:
I don't recall why we added this, but parity with DD seems likely. I would actually prefer to add an Event with the `request_start_time` - that's more meaningful than a "fake span" and is similar to what we do with Sidekiq job instrumentation with `created_at` and `enqueued_at` https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/57cabe3fac44a0921af542b62dc34ef285104a5c/instrumentation/sidekiq/lib/opentelemetry/instrumentation/sidekiq/middlewares/server/tracer_middleware.rb#L35-L36

In general, I am not fond of encoding timestamps in attributes (as integers or strings) and then parsing them on the backend. This sort of thing is exactly what Events are intended for.

Reply:
We did discuss recording it as an event but we didn't know what it would mean for a span event to have a timestamp that was _earlier_ than start timestamp of the event, so I thought it was best not to add it at all.

If users choose to add it to the allow list still that gives them access to the header without us having to add special logic.

Reply:
There is ample precedent for an event timestamp earlier than the span start time - in particular, see the Sidekiq examples I shared. I strongly believe this should be an event rather than an attribute or a span. Requiring backends to parse and interpret stringly-typed attributes is bad form IMO, especially when we have a well structured and typed alternative.

Reply:
Fine to drop and fine to restructure as an event. It was a thing that came from dd and wasn’t even standard across dd sdks, it predated the open tracing Nginx instrumentation iirc and was added as a one off work around for a specific user, although over time it did grow to become a somewhat popular feature with ddtracerb users. I believe the correct approach is to “simply” instrument your proxy.

Reply:
I have moved it into an event.